docs: Another approach for combining Expressions & Parentheses (#2460)
PrQL/prql: web/book/src/syntax/expressions-and-operators.md

## Operator precedence

This table shows operator precedence. Use parentheses `()` to prioritize
operations and for function calls (see below.) _Question: Are parentheses

> operators? Should they be in this table?_

Reviewer: Yes, maybe they should be first?
@aljazerzen ?

Reviewer 2: I'd say they are not, but it doesn't hurt to include them. And yes, they should be first, yes.

Author: `()` is now listed at the top. Precedence is listed as "0" - I'd look for guidance here about whether to set its priority to zero, or to set it to 1 and increase all the others.

<!-- markdownlint-disable MD033 — the `|` characters need to be escaped, and surrounded with tags rather than backticks   -->

|          Group | Operators         | Precedence | Associativity |
| -------------: | ----------------- | :--------: | :-----------: |
| identifier dot | `.`               |     1      |               |
|          unary | `- + ! ==`        |     2      |               |
|          range | `..`              |     3      |               |
|            mul | `* / %`           |     4      | left-to-right |
|            add | `+ -`             |     5      | left-to-right |
|        compare | `== != <= >= < >` |     6      | left-to-right |
|       coalesce | `??`              |     7      | left-to-right |
|            and | `&&`              |     8      | left-to-right |
|             or | <code>\|\|</code> |     9      | left-to-right |
|  function call |                   |     10     |               |
|   _parens (?)_ | `()`              |     ??     |   see below   |

## Parentheses

PRQL uses parentheses `()` for several purposes:

- Parentheses group operands to control the order of evaluation, for example:
  `((1 + x) * y)`

> - Parentheses set off an [inner transform](./inner-transforms.md) for the

Reviewer: Is the "set off" term clear? I'm not familiar with it.

How about just "are required"?

Author: If you're asking about it, I guess "set off" is not clear :-) And I don't mean like a firecracker...

I've changed it to "denote" to see if that reads better.

  `group ()` and `window ()` transforms. _Question: Are there any other
  transforms that use parentheses like this?_

- Parentheses set off a minus sign for a function argument, for example:
  `add (-1) (-3)`

- Parentheses set off nested function calls that contain a pipe, either the `|`
  symbol or a new line. “Nested” means within a transform; i.e. not just the
  main pipeline, for example: `(column-name | in 0..20)`

- Parentheses wrap a function call that is part of a larger expression, for

>   example: `round 0 (sum distance)` _Question: Is it necessary to state "on the

Reviewer: IIUC this is the same point as https://github.com/PRQL/prql/pull/2460/files#r1168000249.

I had this as a footnote so that it was complete. Having the semantics fully listed somewhere means there is some canonical source of truth, which is important if we want to change it. But I agree it's not really relevant for most users — we could still have it in the footnote, or link to a fuller description in the code / in some tests...

Author: see the change that I just committed re: "on the right-hand side..."

  right-hand side of an assignment"? Are there counterexamples?_

>   _Question: Why not require parentheses around function calls in a list? Why

Reviewer: I def see that perspective. But it would add some syntactic noise to:

```elm
aggregate [
  max salary, # would be invalid
  min salary, # would be invalid
  count,
]
```

I'd not set either way, though not in a rush to change from the current state

Author: OK. There _are_ counterexamples. I'll include the "on the right hand side..."

  preserve this (confusing) special case?_

> - Parentheses are not required for expressions that do not contain function

Reviewer: 👍

  calls, for example: `foo + bar`.

Here's a set of examples of these rules:

> ```prql no-fmt

Reviewer: Can I confirm there are no changes to this section?

(It can be easier to split the PRs up so the file move is in a different PR to the text changes. But I recognize that can also add overhead to the process...)

Author: The only change to the example is that I changed "it's contains a pipe" to "it contains a pipe"

(Someday I need to find how to regenerate the snapshots of SQL. I currently use `task test-rust`but that sometimes takes 15 minutes in the Dev Container if crates need to be compiled again.)

from employees
# Requires parentheses, because it contains a pipe
derive is_proximate = (distance | in 0..20)
# Requires parentheses, because it's a function call
derive total_distance = (sum distance)
# `??` doesn't require parentheses, as it's not a function call
derive min_capped_distance = (min distance ?? 5)
# No parentheses needed, because no function call
derive travel_time = distance / 40
# No inner parentheses needed around `1+1` because no function call
derive distance_rounded_2_dp = (round 1+1 distance)
derive [
  # Requires parentheses, because it contains a pipe
  is_far = (distance | in 100..),
  # The left value of the range requires parentheses,
  # because of the minus sign
  is_negative = (distance | in (-100..0)),
  # ...this is equivalent
  is_negative = (distance | in (-100)..0),
  # Doesn't require parentheses, because it's in a list (confusing, see footnote)!
  average_distance = average distance,
]
# Requires parentheses because of the minus sign
sort (-distance)
# A list is fine too
sort [-distance]
```

> ```admonish note

Reviewer: 👍

## Operator precedence

This table shows operator precedence. Use parentheses `()` to prioritize
operations and for function calls (see the discussion below.)

<!-- markdownlint-disable MD033 — the `|` characters need to be escaped, and surrounded with tags rather than backticks   -->

|          Group | Operators         | Precedence | Associativity |
| -------------: | ----------------- | :--------: | :-----------: |

> |    parentheses | `()`              |     0      |   see below   |

Reviewer: @aljazerzen for your post-merge review. Not really an operator but I thought it did belong in the precedence table, let us know if not

Reviewer 2: I replied about that somewhere - it's fine even if they are technically not operators. , understandable enough.

| identifier dot | `.`               |     1      |               |
|          unary | `- + ! ==`        |     2      |               |
|          range | `..`              |     3      |               |
|            mul | `* / %`           |     4      | left-to-right |
|            add | `+ -`             |     5      | left-to-right |
|        compare | `== != <= >= < >` |     6      | left-to-right |
|       coalesce | `??`              |     7      | left-to-right |
|            and | `&&`              |     8      | left-to-right |
|             or | <code>\|\|</code> |     9      | left-to-right |
|  function call |                   |     10     |               |

## Parentheses

> PRQL uses parentheses `()` for several purposes:

Reviewer: I think this reads quite well now! Thanks @richb-hanover !

- Parentheses group operands to control the order of evaluation, for example:
  `((1 + x) * y)`

> - Parentheses denote an [inner transform](./inner-transforms.md) for the

Reviewer: In a separate PR, we could merge the `inner transform` page into this, it's a bit sparse and lonely on its own...

Author: I wondered about merging it in. I guess `group` and `window` can link to the sub-page.

  `group ()` and `window ()` transforms.

> - Parentheses denote a minus sign for a function argument, for example:

Reviewer: ```suggestion
- Parentheses denote a minus sign of a function argument, for example:
```
of / for?

Author: of

derive is_proximate = (distance | in 0..20)
# Requires parentheses, because it's a function call
derive total_distance = (sum distance)
# `??` doesn't require parentheses, as it's not a function call
derive min_capped_distance = (min distance ?? 5)
# No parentheses needed, because no function call
derive travel_time = distance / 40
# No inner parentheses needed around `1+1` because no function call
derive distance_rounded_2_dp = (round 1+1 distance)
derive [
  # Requires parentheses, because it contains a pipe
  is_far = (distance | in 100..),
  # The left value of the range requires parentheses,
  # because of the minus sign
  is_negative = (distance | in (-100..0)),
  # ...this is equivalent
  is_negative = (distance | in (-100)..0),
  # Doesn't require parentheses, because it's in a list (confusing, see footnote)!
  average_distance = average distance,
]
# Requires parentheses because of the minus sign
sort (-distance)
# A list is fine too
sort [-distance]
```

Note: The `total_distance` statement below generates an error because the
function is not in a list. (The PRQL compiler should display a better error
message.)

```prql error
from employees

> derive total_distance = sum distance # generates the error shown below

Reviewer: 👍

derive other_distance = (sum distance) # works as expected
```

```admonish note

> We are seeking feedback about these rules to see if they can be more intuitive.

Reviewer: ```suggestion
We're continuing to think whether these rules can be more intuitive.
```

(I actually think this is one place where we're not really "seeking feedback" (one of the **only** places) — we know it's over-complicated but need to think how it all fits together. We're seeking genius ideas if someone has them, but it would require lots of context rather than just "this part is awkward")

Author: Agreed.

- Parentheses group operands to control the order of evaluation, for example:
  `((1 + x) * y)`

- Parentheses denote an [inner transform](./inner-transforms.md) for the
  `group ()` and `window ()` transforms. 

> - Parentheses denote a minus sign for a function argument, for example:

Reviewer: `denote` defn: "be a sign of; indicate."

I don't think that's quite right....

"delimit"? Kinda formal / clunky but the meaning is closer...

Author: Yeah, clunky and formal. But let's go with it.

sort [-distance]
```

Note: The `total_distance` statement below generates an error because the function is not in a list. 
(The PRQL compiler should display a better error message.)

```prql error
from employees
derive total_distance = sum distance # generates the error shown below
derive other_distance = (sum distance) # works as expected
```

```admonish note
We are seeking feedback about these rules to see if they can be more intuitive.
We're also planning to make the error messages much better,

> so the compiler can help out.

Reviewer: 👍
